File: FrontEnd/src/app/domain/trabajo/trabajo.service.ts
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { TrabajoRepository } from './trabajo.repository';
import { Trabajo, TrabajoResponse, FinalizarTrabajoRequest, TrabajoClienteResponse } from './trabajo.model';

@Injectable({ providedIn: 'root' })
export class TrabajoService {
  private readonly repository = inject(TrabajoRepository);

  crearTrabajo(idSolicitud: number): Observable<Trabajo> {
    return this.repository.crearTrabajo(idSolicitud);
  }

  iniciarTrabajo(idTrabajo: number): Observable<TrabajoResponse> {
    return this.repository.iniciarTrabajo(idTrabajo);
  }

  pausarTrabajo(idTrabajo: number): Observable<TrabajoResponse> {
    return this.repository.pausarTrabajo(idTrabajo);
  }

  reanudarTrabajo(idTrabajo: number): Observable<TrabajoResponse> {
    return this.repository.reanudarTrabajo(idTrabajo);
  }

  finalizarTrabajo(idTrabajo: number, observaciones: string, montoFinal: number, duracionReal?: number, montoAdicional?: number, descripcionAdicional?: string, fotoTrabajo?: string): Observable<TrabajoResponse> {
    const request: FinalizarTrabajoRequest = {
      observaciones,
      montoFinal,
      duracionReal: duracionReal || null,
      montoAdicional: montoAdicional || null,
      descripcionAdicional: descripcionAdicional || null,
      fotoTrabajo: fotoTrabajo || null
    };
    return this.repository.finalizarTrabajo(idTrabajo, request);
  }

  cancelarTrabajo(idTrabajo: number, motivo: string): Observable<TrabajoResponse> {
    return this.repository.cancelarTrabajo(idTrabajo, motivo);
  }

  obtenerTrabajo(idTrabajo: number): Observable<TrabajoResponse> {
    return this.repository.obtenerTrabajo(idTrabajo);
  }

  obtenerTrabajoPorSolicitud(idSolicitud: number): Observable<TrabajoResponse> {
    return this.repository.obtenerTrabajoPorSolicitud(idSolicitud);
  }

  obtenerTrabajosPorProfesional(idProfesional: number, estado?: string): Observable<TrabajoResponse[]> {
    return this.repository.obtenerTrabajosPorProfesional(idProfesional, estado);
  }

  obtenerTrabajosPorUsuario(idUsuario: number, estado?: string): Observable<TrabajoResponse[]> {
    return this.repository.obtenerTrabajosPorUsuario(idUsuario, estado);
  }

  obtenerTrabajosSinFactura(): Observable<TrabajoResponse[]> {
    return this.repository.obtenerTrabajosSinFactura();
  }

  obtenerTrabajosFinalizadosPorCliente(idUsuario: number): Observable<TrabajoClienteResponse[]> {
    return this.repository.obtenerTrabajosFinalizadosPorCliente(idUsuario);
  }
}

File: BackEnd/App/src/main/java/ar/edu/utn/frc/tup/app/services/TrabajoService.java
package ar.edu.utn.frc.tup.app.services;

import ar.edu.utn.frc.tup.app.dtos.request.trabajo.FinalizarTrabajoRequest;
import ar.edu.utn.frc.tup.app.dtos.response.trabajo.TrabajoCanceladoNotificacionDTO;
import ar.edu.utn.frc.tup.app.dtos.response.trabajo.TrabajoClienteResponse;
import ar.edu.utn.frc.tup.app.dtos.response.trabajo.TrabajoFinalizadoNotificacionDTO;
import ar.edu.utn.frc.tup.app.dtos.response.trabajo.TrabajoResponse;
import ar.edu.utn.frc.tup.app.entities.Factura;
import ar.edu.utn.frc.tup.app.entities.Trabajo;

import java.util.List;

public interface TrabajoService {
    Trabajo crearTrabajo(Integer idSolicitud);
    TrabajoResponse iniciarTrabajo(Integer idTrabajo);
    TrabajoResponse pausarTrabajo(Integer idTrabajo);
    TrabajoResponse reanudarTrabajo(Integer idTrabajo);
    TrabajoResponse finalizarTrabajo(Integer idTrabajo, FinalizarTrabajoRequest request);
    TrabajoResponse cancelarTrabajo(Integer idTrabajo, String motivoCancelacion);
    TrabajoResponse obtenerTrabajoPorId(Integer idTrabajo);
    TrabajoResponse obtenerTrabajoPorSolicitud(Integer idSolicitud);
    List<TrabajoResponse> obtenerTrabajosPorProfesional(Integer idProfesional);
    List<TrabajoResponse> obtenerTrabajosPorProfesionalyEstado(Integer idProfesional, String estado);
    List<TrabajoResponse> obtenerTrabajosPorUsuario(Integer idUsuario, String estado);
    List<TrabajoResponse> obtenerTrabajosSinFactura();
    List<TrabajoClienteResponse> obtenerTrabajosFinalizadosPorCliente(Integer idUsuario);
    List<TrabajoCanceladoNotificacionDTO> obtenerTrabajosCanceladosPorCliente(Integer idUsuario);
    List<TrabajoFinalizadoNotificacionDTO> obtenerTrabajosFinalizadosParaNotificar(Integer idUsuario);
}
